solidity/contracts/UpgradableProxyStorage.sol

```diff
@@ -0,0 +1,206 @@
+
+    /// @dev Storage slot with the upgrade time delay. Upgrade time delay defines a
+    /// period for implementation upgrade.
+    /// This is the keccak-256 hash of "network.keep.proxy.upgradetimedelay"
+    /// subtracted by 1, and is validated in the constructor.
+    bytes32 internal constant UPGRADE_TIME_DELAY_SLOT = 0xa0361fd71e28dd9e0781644db9ca971cf5aeafd12b5d0aaaebff68299a2cbbee;
+
+    /// @dev Storage slot with the ID of the current implementation. The ID is
+    /// an unique identifier of the implementation calculated based on the version.
+    /// This is the keccak-256 hash of "network.keep.proxy.currentimplementationid"
+    /// subtracted by 1, and is validated in the constructor.
+    bytes32 internal constant CURRENT_IMPL_ID_SLOT = 0x2fdc2b7761cb35b3c6cce7021b2c04230161423bd5b9520fd8a7ae15aa03ac9f;
+
+    /// @dev Storage slot with the new implementation ID.
+    /// This is the keccak-256 hash of "network.keep.proxy.upgradeimplementationid"
+    /// subtracted by 1, and is validated in the constructor.
+    bytes32 internal constant UPGRADE_IMPL_ID_SLOT = 0x1836489d800664edd6c29124d8cf32e481082cc2687a6f957a01e0f20a003d40;
+
+    /// @dev Storage slot with the implementation address upgrade initiation.
+    /// This is the keccak-256 hash of "network.keep.proxy.upgradeinitiatedtimestamp"
+    /// subtracted by 1, and is validated in the constructor.
+    bytes32 internal constant UPGRADE_INIT_TIMESTAMP_SLOT = 0xf4917e72ec7208070be0cec8873a121e5971a5d90b452600ca1524183a489ed5;
+
+    // Structure holding details of an implementation.
+    struct Implementation {
+        string version;
+        address implementationContract;
+        bytes initializationData;
+    }
+
+    // Mapping from version ID to implementation contract. It is expected that
+    // the map key is a keccak256 of the implementation version.
+    mapping(uint256 => Implementation) public implementations;
+
+    constructor() public {
+        assertSlot(ADMIN_SLOT, "eip1967.proxy.admin");
+        assertSlot(
+            UPGRADE_TIME_DELAY_SLOT,
+            "network.keep.proxy.upgradetimedelay"
+        );
+        assertSlot(
+            CURRENT_IMPL_ID_SLOT,
+            "network.keep.proxy.currentimplementationid"
+        );
+        assertSlot(
+            UPGRADE_IMPL_ID_SLOT,
+            "network.keep.proxy.upgradeimplementationid"
+        );
+        assertSlot(
+            UPGRADE_INIT_TIMESTAMP_SLOT,
+            "network.keep.proxy.upgradeinitiatedtimestamp"
+        );
+    }
+
+    function getImplementation(string memory _version)
+        internal
+        view
+        returns (Implementation memory)
+    {
+        return
+            getImplementation(uint256(keccak256(abi.encodePacked(_version))));
+    }
+
+    function getImplementation(uint256 _version)
+        internal
+        view
+        returns (Implementation memory)
+    {
+        return implementations[_version];
+    }
+
+    function setImplementation(
```

Comment:
Thoughts on naming it `addImplementation`? We are storing all implementations in the map, we are not overwriting anything.

EDIT: I think we shouldn't store implementations in this mapping. Everyone can track the history of upgrades from the events we are emitting. Storing just an initialization data we are committing to should be enough. And we can even delete the data from this mapping once the initialization is done.